segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/LogReader.java

```diff
@@ -168,11 +169,10 @@ private void openNextLedger(LedgerAddress address) throws DurableDataLogExceptio
         @Getter
         private final LedgerAddress address;
 
-        @SneakyThrows(IOException.class)
-        ReadItem(LedgerEntry entry, LedgerMetadata ledgerMetadata) {
-            this.address = new LedgerAddress(ledgerMetadata, entry.getEntryId());
-            this.payload = entry.getEntryInputStream();
-            this.length = this.payload.available();
+        ReadItem(long entryId, byte[] entry, LedgerMetadata ledgerMetadata) {
```

Comment:
This will allocate a new Java byte array from the internal `ByteBuffer`. The reason it used `getEntryInputStream` was precisely to avoid that.

Let's figure out a way to get an `InputStream` directly from the underlying buffer. You can use `ByteBufInputStream`, but I am sad to see this API gone.

Reply:
Ok I see below the reason why you did it this way. Then we should take care to release this buffer when `getNext` is invoked (for the previously returned item) or when we close the `LogReader` class.

In any case I want to avoid having to allocate too many Java byte arrays, which will make the recovery process unnecessarily slow.

Reply:
@andreipaduroio
I am performing the copy because I am not sure about the contract of ReadItem.

Will getInputStream be called once and closed once?

We can release the buffer when the caller Invokes InputStream#close and close all unconsumed entries in LogReader#close.

If I understand correctly you are telling me that this approach is doable and we won't introduce memory leaks.

I will try to update the patch soon.

I have to check more but as the prev BK API did not provide a close() method the copy was performed inside BK

Reply:
@andreipaduroiu done